### PR TITLE
Manual & Scheduled dynamic flask templates, filtering module

### PR DESCRIPTION
**Pull request by c0sc0**
hahaha PRs go brrrrrrrrrrrrrr (i think i finally got this shit figured out l0l, sorry for holding up the line (no more committing straight to master 👍  ))

## Dynamic web interface

- grab open & close as well, using it to calculate premarket & total moves. Also the close for the current day is the current price even with the 1d interval ;) (this is where weighting/distribution fitting would come into play)

- fully (at least on my desktop, iPhone & iPad) responsive templates. Dynamic.html has a form for scanning params & a refresh button, dynamicRefresh.html has no form/button and reloads the page every minute. More on that later. (I'd like to use AJAX to update the table instead of forcing a refresh)

- globals in market_scanner moved into the class & constructor so you can change up the params in 1 line & easily import the module in other scripts

- double the number of worker processes to deal with the yfinance bottleneck

- flask_server is kind of a mess right now. But it's a hot mess. I'm trying to wrap all of it into a function (so it can be put onto nginx) but I didn't get all of the way there.

- There's some ugly validation in the /reload/ endpoint (this is POSTed with the form deets in the manual template)

- To switch between the manual/automatic templates, change the `AUTO_REFRESH = True` line

- The update frequency can be set (in minutes) by changing `UPDATE_FREQ = 5` 

- The scheduler could probably use some work. It spawns a timer every `UPDATE_FREQ` minutes, which spawns a thread which runs the scanner. It does not check if something went wrong, nor does it wait before the previous job is finished before it starts a new onwe

## filtering module

- filters the tickerlist on some criterion, splits it into include/exclude CSVs

- have market_scanner load the tickerlist by passing a `doFilter` boolean to main_func()

- e.g. `results = mainObj(_month_cuttoff=6,_day_cuttoff=3,_std_cuttoff=9).main_func(doFilter=False)`

I think that's it. I probably missed some modules in requirements.txt

### CODE REVIEW

**Comment by c0sc0** [comment]
You might notice that `len(filtered_tickers + excluded_tickers)` is less than the total number of tickers. I assume this is because yfinance doesnt have data for those tickers, in which case we should "pre-filter" up to the 100th percentile so we don't waste precious API calls

File: flask_server.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        month_cut = int(request.form.get('month_cutoff'))
+        #didn't want to do this in JS lmao
+        #The forms have limits, but users can type in whatever number
+        #It's me, yandaredev
+        if month_cut > 12:
+          month_cut = 12
+        if month_cut < 1:
+          month_cut = 1
+        day_cut = int(request.form.get('day_cutoff'))
+        if day_cut > 30:
```
Comment (elizabethshirley) [inline flask_server.py]: bruh... what happens in January, March, May, July, August, October, December?

File: flask_server.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      return render_template('dynamic.html', stonks=stonks, month_cut=6, day_cut=3, std_cut=2, update_time=stonk_search.updated_at)
+
+  @app.route('/reload/', endpoint="reload", methods=["POST"]) 
+  def refresh():
+      if request.method == "POST":
+        month_cut = int(request.form.get('month_cutoff'))
+        #didn't want to do this in JS lmao
+        #The forms have limits, but users can type in whatever number
+        #It's me, yandaredev
+        if month_cut > 12:
```
Comment (elizabethshirley) [inline flask_server.py]: month_cut = max(min(month_cut, 12),1)

or 

def clamp(n, smallest, largest): return max(smallest, min(n, largest))

month_cut = clamp(month_cut, 1,12)

**Review by elizabethshirley — CHANGES_REQUESTED** [review]
Clean up the flask_server with my minor changes, and you're ready

File: flask_server.py (hunk 1 above)
Comment (c0sc0) [reply flask_server.py]: ahahahahah good catch

coding @ 4am make my brain not work good

File: flask_server.py (hunk 2 above)
Comment (c0sc0) [reply flask_server.py]: [It's gonna be clamp this, clamp that. Bada-climp, bada-clamp!](https://i.imgur.com/rnrezxa.jpg)

File: flask_server.py (hunk 1 above)
Comment (elizabethshirley) [reply flask_server.py]: Probably want to use a date validation library, like dateutil.parser. Things get complicated, especially with leap years

**Comment by c0sc0** [comment]
@SamPom100 @alexshirley fixes are locked & loaded!

**Comment by c0sc0** [comment]
(I thought Sam had the dynamic site hosted somewhere but I now see that it's a frozen version and i was trying to be sneaky for nothing 😆 )

But yeah. I didn't disable the /reload/ endpoint when auto_refresh is on, thus some asshole could keep POSTing to it and it would keep firing up market_scanners & lock up your server. Though really they should be split into 2 different flask scripts so stuff like this doesn't happen.

Fixes are even more locked and loaded. Emphasis on the 'locked' part

**Comment by SamPom100** [comment]
very nice work, looks great. I added the dynamic version as a second page to the website.

**Comment by c0sc0** [comment]
Very nice!

I might put the auto-refreshing version on a Hetzner box or something if people are interested

**Pull request closed without merging**